### ISSUE TITLE
Dsolve clobbers integration constants

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#15999]
If an ODE contains symbols like C1 then `nth_algebraic` will return an incorrect solution:
```julia
In [1]: C1 = Symbol('C1')                                                                                                                                     

In [2]: dsolve(f(x).diff(x)-C1)                                                                                                                               
Out[2]: f(x) = C₁⋅(x + 1)
```
This solution should contain a new integration constant e.g. C2.

**Comment by oscarbenjamin** [comment]
Actually this is not just nth_algebraic. It also happens with nth_linear_constant_coeff_homogeneous. Maybe this is a general problem with dsolve:
```julia
In [18]: eq = f(x).diff(x) + C1*f(x)                                                                                                                          

In [19]: eq                                                                                                                                                   
Out[19]: 
          d       
C₁⋅f(x) + ──(f(x))
          dx      

In [20]: dsolve(eq)                                                                                                                                           
Out[20]: 
           -C₂⋅x
f(x) = C₁⋅ℯ  
```
Here the C1 and C1 are the wrong way round.

**Comment by smichr** [comment]
So initial constants should be replaced with Dummies and returned constants be renumbered to avoid the clash and then the original constants returned into place?

**Comment by asmeurer** [comment]
Duplicate of #4838

**Comment by oscarbenjamin** [comment]
I'm going to reopen this because it's only a partial duplicate of #4838.

The first and second example I referred to are showing different overlapping problems. The problem in the first example is to do with the nth_algebraic solver (which I added myself) not using the `get_numbered_constants` function:
[https://github.com/sympy/sympy/blob/a22a512ca2330053eb2b0185aea92cd7f05677ea/sympy/solvers/ode.py#L371](https://github.com/sympy/sympy/blob/a22a512ca2330053eb2b0185aea92cd7f05677ea/sympy/solvers/ode.py#L371)
[https://github.com/sympy/sympy/blob/a22a512ca2330053eb2b0185aea92cd7f05677ea/sympy/solvers/ode.py#L4001](https://github.com/sympy/sympy/blob/a22a512ca2330053eb2b0185aea92cd7f05677ea/sympy/solvers/ode.py#L4001)
If nth_algebraic was using `get_numbered_constants` then it would introduce a new constant `C2` to the solution since it would check that `C1` was in the ODE first. I think that the reason I didn't do that at the time is that `get_numbered_constants` returns a fixed size list and I had no way of knowing how many constants I would need since mine would be called inside `solve` somewhere. I'll send a PR adding a generator version of `get_numbered_constants` and fixing `nth_algebraic`. That can then close this issue.

The other issue as shown in the second example is that `dsolve` simplifies constants in the solution disrespecting those that were in the original ODE - that's a duplicate of #4838 for which there is already a resurrectable PR #13752.

Looking through #4838 and related issues and @smichr's comment above I wonder if the real solution is that actually integration constants should always be `Dummy` symbols not just internally to `dsolve` but in the returned solution as well. This should go into a third issue but for example:
```julia
In [9]: sol1 = dsolve(f(x).diff(x)-x).rhs                                                                                                                     

In [10]: sol2 = dsolve(f(x).diff(x)-x**2).rhs                                                                                                                 

In [11]: sol1                                                                                                                                                 
Out[11]: 
      2
     x 
C₁ + ──
     2 

In [12]: sol2                                                                                                                                                 
Out[12]: 
      3
     x 
C₁ + ──
     3 

In [13]: sol1 - sol2                                                                                                                                          
Out[13]: 
   3    2
  x    x 
- ── + ──
  3    2 
```
Those two integration constants should not cancel in the subtraction.
